### PR TITLE
LVNOD-89-Gift-logic

### PR DESCRIPTION
**Pull request by denisbodnar**
## JIRA

* [Main JIRA ticket](https://ssu-jira.softserveinc.com/browse/LVNOD-XXX)
* [JIRA subtask](https://ssu-jira.softserveinc.com/browse/LVNOD-XXX)

## Code reviewers

- [x] @yarvanya
- [x] @yurii-kokhalevych
- [x] @yuliannakorol

### Second Level Review

- [ ] @lhalam
- [ ] @4asof4ik

### CODE REVIEW

**Review by yarvanya — DISMISSED** [review]

**Review by yurii-kokhalevych — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

**Review by yurii-kokhalevych — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

File: server/controllers/gift.controller.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  create(req, res) {
+    let assignGift = Object.assign({}, req.body, {event_id: req.params.id});
+
+    Gift.create(assignGift)
+    .then(gift => res.status(201).send(gift))
+    .catch(error => res.status(400).send(error));
+  },
+
+  list(req, res) {
+    console.log(req.params);
```
Comment (4asof4ik) [inline server/controllers/gift.controller.js]: ?

File: server/controllers/gift.controller.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+  retrieve(req, res) {
+    Gift.findById(
+      req.params.giftid, {
+      attributes: [
+          'name', 'description', 'link', 'is_available', 'event_id'
+      ]})
+      .then(gift => {
+        if (gift.dataValues.event_id !== Number(req.params.id) || !gift)  {
+          return res.status(400).send(messages.giftNotFound);
```
Comment (4asof4ik) [inline server/controllers/gift.controller.js]: 404 but not 400

File: server/controllers/gift.controller.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        }
+      })
+      .catch(error => {
+        return res.status(400).send(error);
+      });
+  },
+
+  update(req, res) {
+    Gift.findById(req.params.giftid).then(gift => {
+      if (gift.dataValues.event_id !== Number(req.params.id) || !gift) {
```
Comment (4asof4ik) [inline server/controllers/gift.controller.js]: create a common function for this validation ( we use here, destroy  and retrieve)

File: server/routes/index.js [hunk 4]
```diff
@@ -24,4 +25,9 @@ module.exports = app => {
   app.post('/api/auth/login', authController.login);
   app.get('/api/user/activation/:token', authController.activation);
 
+  app.post('/api/event/:id/gifts', giftController.create);
+  app.get('/api/event/:id/gifts', giftController.list);
+  app.get('/api/event/:id/gift/:giftid', giftController.retrieve);
```
Comment (4asof4ik) [inline server/routes/index.js]: gift_id

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

**Review by yarvanya — DISMISSED** [review]

File: server/controllers/gift.controller.js [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  return Guest.findOne({
+    where: {event_id: Number(eventId),
+      $and: {user_id: userId}
+    }
+  })
+  .then(result => result ? result.dataValues.user_id : false)
+  .catch(error => error);
+};
+
+const eventIsDraft = (eventId) => {
```
Comment (yurii-kokhalevych) [inline server/controllers/gift.controller.js]: write without ()  - eventId

**Review by yurii-kokhalevych — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by yurii-kokhalevych — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

File: server/controllers/gift.controller.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
+'use strict';
+
+const Gift = require('../../config/db').Gift;
```
Comment (omaik) [inline server/controllers/gift.controller.js]: Could you do something like following instead?
```js
const DB = require('../../config/db')
const Gift = DB.Gift
....
```
Now you requiring the same file 3 times which looks redundant.

File: server/controllers/gift.controller.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+      $and: {status_event_id: 1}
+    }
+  })
+  .then(result => result && Number(result.dataValues.id))
+  .catch(error => error);
+};
+
+module.exports = {
+  create(req, res) {
+    let assignGift = Object.assign({}, req.body, {event_id: req.params.id});
```
Comment (omaik) [inline server/controllers/gift.controller.js]: better name would be `giftParams`

**Review by yarvanya — DISMISSED** [review]

**Review by yurii-kokhalevych — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

File: server/routes/index.js [hunk 8]
```diff
@@ -27,4 +28,10 @@ module.exports = app => {
 
   app.post('/api/event/:id/guest/invite', guestController.invite);
   app.get('/api/event/:id/guest/get', guestController.list);
+
+  app.post('/api/event/:id/gifts', giftController.create);
+  app.get('/api/event/:id/gifts', giftController.list);
```
Comment (4asof4ik) [inline server/routes/index.js]: events 32 and 33

File: server/controllers/gift.controller.js [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+  })
+  .then(result => result && Number(result.dataValues.id))
+  .catch(error => error);
+};
+
+module.exports = {
+  create(req, res) {
+    let giftParams = Object.assign({}, req.body, {event_id: req.params.id});
+    Gift.create(giftParams)
+    .then(gift => isEventOwner(req.params.id, req.decoded.id, gift)
```
Comment (4asof4ik) [inline server/controllers/gift.controller.js]: the only owner can create a gift, so why check that?

File: server/controllers/gift.controller.js [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    )
+    .catch(error => res.status(400).send(error));
+  },
+
+  list(req, res) {
+    Gift.findAll({
+      attributes: ['id', 'name', 'description', 'link', 'is_available', 'status'],
+      where: {event_id: req.params.id}
+    })
+    .then(gift => isEventOwner(req.params.id, req.decoded.id, gift)
```
Comment (4asof4ik) [inline server/controllers/gift.controller.js]: gifts

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

File: server/controllers/gift.controller.js (hunk 9 above)
Comment (denisbodnar) [reply server/controllers/gift.controller.js]: without this check any logged in user can create any gift in any event behind the UI

File: server/routes/index.js (hunk 8 above)
Comment (denisbodnar) [reply server/routes/index.js]: but there next we have an id of that event

**Review by yarvanya — DISMISSED** [review]

**Review by yurii-kokhalevych — DISMISSED** [review]

**Review by yurii-kokhalevych — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by yurii-kokhalevych — APPROVED** [review]

**Review by yuliannakorol — APPROVED** [review]

**Review by yarvanya — APPROVED** [review]

**Review by 4asof4ik — APPROVED** [review]

**Pull request merged**